libnvme/src/firmware.rs

```diff
@@ -231,34 +231,31 @@ impl<'ctrl> WriteLockedController<'ctrl> {
     ) -> Result<(), FirmwareLoadError> {
         // TODO swap to the libnvme granularity function Andy is adding
         const CHUNK_SIZE: u64 = 0x1000;
-        let mut offset = 0u64;
-
-        loop {
-            let len =
-                CHUNK_SIZE.try_into().expect("32-bit systems unsupported");
-            let mut buf = vec![0; len];
-            match data.read_exact(&mut buf) {
-                Ok(_) => self.firmware_load_chunk(buf.as_slice(), offset)?,
-                Err(e) if e.kind() == ErrorKind::Interrupted => continue,
-                Err(e) if e.kind() == ErrorKind::UnexpectedEof => {
-                    let mut buf = Vec::new();
-                    data.read_to_end(&mut buf)?;
-                    if buf.is_empty() {
-                        break;
-                    }
-                    buf.resize(len, 0);
-                    self.firmware_load_chunk(buf.as_slice(), offset)?;
-                    break;
-                }
-                Err(e) => return Err(e.into()),
-            }
 
+        let size = CHUNK_SIZE.try_into().expect("32-bit systems unsupported");
+        let mut offset = 0u64;
+        let mut buf = Vec::new();
+
+        // Firmware blobs tend to be a few MB in size. For simplicity we are
+        // going to read everything passed to us via the reader into a Vec so
+        // that we split it into proper chunk sizes.
+        data.read_to_end(&mut buf)?;
```

Comment:
This looks good to me. Re: your note about adding a `take` to prevent a caller passing something huge - should we just take a `&[u8]` instead of a reader if we're going to read the whole thing into memory anyway? It doesn't seem unreasonable - it's just a couple lines of code for a caller to do what you have here if they have a reader, and if they already have the blob in memory for some other reason it's slightly more ergonomic to take the slice directly.